### ISSUE TITLE
Function this type is not checked when assigning function types

### ISSUE THREAD

**Issue by d07RiV** [microsoft/TypeScript#54346]
# Bug Report

### 🔎 Search Terms

`this`

(it seems that search engines ignore "this" keyword completely)

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and documentation does not mention anything about it

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiD29oG8BQ1oHMBOBTXALgBQCUK6G0w8AdhPCLgHQjyZEDkAEriK9AHV42EABMAhBxIUAvqjmoAZgFcawAgEta0RYlir1WmkQIALDRABccRGTQYzFpjnzFpC6nQI6D10tABeAD5oADd4DVFAnT0DTVolA1JUIA)

### 💻 Code

```ts
class Foo {
  greet() {
    console.log('Hello World!')
  }
}

function fooFunction(this: Foo) {
  this.greet()
}

const func: () => void = fooFunction
func()
```

### 🙁 Actual behavior

Type `(this: Foo) => void` should not be assignable to `() => void`

### 🙂 Expected behavior

TypeScript happily compiles this code and it crashes when executed.

**Comment by fatcerberus** [comment]
> The problem is that I can pass functions relying on this as arbitrary callbacks. TypeScript fails to warn me about these scenarios.

Yes, I understand.  You can type your callbacks as `(this: void) => void`, which will disallow the assignment from, e.g. `fooFunction` above.  The problem is that `() => void` is implicitly treated as `(this: any) => void`.

**Comment by jcalz** [comment]
Duplicate of #7968
